Attribute Analysis.ipynb

```diff
@@ -602,6 +606,379 @@
+    "        return\n",
+    "    else:\n",
+    "        plt.show()\n",
+    "\n",
+    "    if len(infinities) > 0:\n",
+    "        print(\"Categories where one attribute had 0 occurrences with that object:\\n\")\n",
+    "        \n",
+    "    # Categories where all attributes had 0 occurrences with that object\n",
+    "    for index in infinities:\n",
+    "        print(\"{0} had 0 occurrences with all values of the attribute\".format(xaxis_sig_topn[index]))\n",
+    "\n",
+    "    print(\"The most common attribute associated each category in the graph is:\\n\")\n",
+    "    for index in reversed(range(topn)):\n",
+    "        #attr = int(max_counts[top_indices][index])\n",
+    "        attr = int(max_c[index])\n",
+    "        if sort_by == 'pvalue':\n",
+    "            instance_sig_top = instance_sig_topn[index]\n",
+    "        elif sort_by == 'ratio':\n",
+    "            instance_sig_top = instance_sig_topn[attr][index]\n",
+    "        print(\"{0}: {1}={2}, proportion of total instances: {3}\\n\".format(xaxis_sig_topn[index], attr_names[attr], math.ceil(instance_sig_top*dataset.num_skin_tone_images[attr]), round(np.absolute(all_ratios_topn[index]), 4)))\n",
+    "        \n",
+    "p_values = np.zeros_like(counts_mat[0])\n",
+    "attributes = np.array([[[-1,-1] for a in range(len(counts_mat[0][0]))] for b in range(len(counts_mat[0]))])\n",
+    "mi = [np.zeros_like(counts_mat[0]) for a in range(num_attrs)]\n",
+    "mi_wilson = [np.zeros_like(counts_mat[0]) for a in range(num_attrs)]\n",
+    "\n",
+    "for i in range(len(categories)):\n",
+    "    for j in range(len(categories)):\n",
+    "        if categories[i] in dataset.people_labels or categories[j] in dataset.people_labels:\n",
+    "            p_values[i][j] = -1\n",
+    "        else:\n",
+    "            min_p = 100\n",
+    "            attr1 = -1\n",
+    "            attr2 = -1\n",
+    "            for a in range(num_attrs):\n",
+    "                for b in range(a+1, num_attrs):\n",
+    "                    len_a = instance_counts[a][j] + instance_counts[a][i] - counts_mat[a][i][j]\n",
+    "                    an = np.zeros(int(len_a))\n",
+    "                    an[:int(counts_mat[a][i][j])] = 1\n",
+    "                    mi[a][i][j] = np.mean(an)\n",
+    "                    mi_wilson[a][i][j] = wilson(np.mean(an), len_a)[0]\n",
+    "                    \n",
+    "                    len_b = instance_counts[b][j] + instance_counts[b][i] - counts_mat[b][i][j]\n",
+    "                    bn = np.zeros(int(len_b))\n",
+    "                    bn[:int(counts_mat[b][i][j])] = 1\n",
+    "                    mi[b][i][j] = np.mean(bn)\n",
+    "                    mi_wilson[b][i][j] = wilson(np.mean(bn), len_b)[0]\n",
+    "                    if mi[a][i][j] > 0 and mi[b][i][j] > 0:\n",
+    "                        p = stats.ttest_ind(an, bn)[1]\n",
+    "                    else:\n",
+    "                        p = 100\n",
```

Comment:
same, not fully sure what's happening here

Reply:
So this is to prevent counts of 0 from showing up on the mutual information feature, because otherwise a lot of the attribute pairs with the lowest p-value, also had one attribute with a value of 0 and I didn't feel like that was very useful to have